pkg/controller/persistentvolume/index.go

```diff
@@ -87,15 +89,29 @@ func (pvIndex *persistentVolumeOrderedIndex) findByClaim(claim *api.PersistentVo
 		unboundVolumes := []*api.PersistentVolume{}
 		for _, volume := range volumes {
 			// volume isn't currently bound or pre-bound.
-			if volume.Spec.ClaimRef == nil {
-				unboundVolumes = append(unboundVolumes, volume)
-				continue
-			}
-
 			if isVolumeBoundToClaim(volume, claim) {
 				// Exact match! No search required.
 				return volume, nil
 			}
+
+			// filter out volumes whose labels don't match the claim's
+			// selector, if specified
+			if volume.Spec.ClaimRef == nil {
+				if claim.Spec.LabelSelector != nil {
+					internalSelector, err := unversioned.LabelSelectorAsSelector(claim.Spec.LabelSelector)
```

Comment:
I'd create the internalSelector only once, outside the loop and then reuse it for matching.